File: deployment_cpp/drone_tracker.cpp
#include "drone_tracker.hpp"

// ============================================================
// IOU CALCULATION
// ============================================================

float calculate_iou(const cv::Vec4f& box1, const cv::Vec4f& box2) {
    float x1 = std::max(box1[0], box2[0]);
    float y1 = std::max(box1[1], box2[1]);
    float x2 = std::min(box1[2], box2[2]);
    float y2 = std::min(box1[3], box2[3]);

    float intersection = std::max(0.0f, x2 - x1) * std::max(0.0f, y2 - y1);
    float area1 = std::max(0.0f, box1[2] - box1[0]) * std::max(0.0f, box1[3] - box1[1]);
    float area2 = std::max(0.0f, box2[2] - box2[0]) * std::max(0.0f, box2[3] - box2[1]);
    float union_area = area1 + area2 - intersection;

    if (union_area <= 0.0f) return 0.0f;
    return intersection / union_area;
}

float calculate_iou(const cv::Vec4i& box1, const cv::Vec4f& box2) {
    cv::Vec4f b1((float)box1[0], (float)box1[1], (float)box1[2], (float)box1[3]);
    return calculate_iou(b1, box2);
}

// ============================================================
// CENTER DISTANCE
// ============================================================

float center_distance(const cv::Vec4i& box1, const cv::Vec4f& box2) {
    float c1x = (box1[0] + box1[2]) / 2.0f;
    float c1y = (box1[1] + box1[3]) / 2.0f;
    float c2x = (box2[0] + box2[2]) / 2.0f;
    float c2y = (box2[1] + box2[3]) / 2.0f;

    float dx = c1x - c2x;
    float dy = c1y - c2y;
    return std::sqrt(dx * dx + dy * dy);
}

// ============================================================
// DRONE TRACK CONSTRUCTOR
// ============================================================

DroneTrack::DroneTrack(const cv::Vec4f& bbox, int track_id)
    : id(track_id),
      x1(bbox[0]), y1(bbox[1]), x2(bbox[2]), y2(bbox[3]),
      age(1), hits(1), missed(0), confirmed(false), confidence(0.0f)
{
    width = std::max(2.0f, x2 - x1);
    height = std::max(2.0f, y2 - y1);
    cx = (x1 + x2) / 2.0f;
    cy = (y1 + y2) / 2.0f;

    kalman.init(6, 4, 0, CV_32F);

    // Transition matrix
    cv::setIdentity(kalman.transitionMatrix);
    kalman.transitionMatrix.at<float>(0, 2) = 1.0f;
    kalman.transitionMatrix.at<float>(1, 3) = 1.0f;

    // Measurement matrix
    kalman.measurementMatrix = cv::Mat::zeros(4, 6, CV_32F);
    kalman.measurementMatrix.at<float>(0, 0) = 1.0f;
    kalman.measurementMatrix.at<float>(1, 1) = 1.0f;
    kalman.measurementMatrix.at<float>(2, 4) = 1.0f;
    kalman.measurementMatrix.at<float>(3, 5) = 1.0f;

    // Process noise
    cv::setIdentity(kalman.processNoiseCov, cv::Scalar::all(0.03f));

    // Measurement noise
    cv::setIdentity(kalman.measurementNoiseCov, cv::Scalar::all(0.15f));

    // Error covariance post
    cv::setIdentity(kalman.errorCovPost, cv::Scalar::all(1.0f));

    // State post
    kalman.statePost.at<float>(0) = cx;
    kalman.statePost.at<float>(1) = cy;
    kalman.statePost.at<float>(2) = 0.0f;
    kalman.statePost.at<float>(3) = 0.0f;
    kalman.statePost.at<float>(4) = width;
    kalman.statePost.at<float>(5) = height;
}

// ============================================================
// KALMAN PREDICT
// ============================================================

cv::Vec4i DroneTrack::predict() {
    cv::Mat prediction = kalman.predict();

    cx = prediction.at<float>(0);
    cy = prediction.at<float>(1);
    width = std::max(2.0f, prediction.at<float>(4));
    height = std::max(2.0f, prediction.at<float>(5));

    x1 = cx - width / 2.0f;
    y1 = cy - height / 2.0f;
    x2 = cx + width / 2.0f;
    y2 = cy + height / 2.0f;

    age++;
    missed++;

    return get_bbox();
}

// ============================================================
// KALMAN UPDATE FROM YOLO
// ============================================================

void DroneTrack::update(const cv::Vec4f& bbox, float conf) {
    float n_x1 = bbox[0];
    float n_y1 = bbox[1];
    float n_x2 = bbox[2];
    float n_y2 = bbox[3];

    float n_cx = (n_x1 + n_x2) / 2.0f;
    float n_cy = (n_y1 + n_y2) / 2.0f;
    float n_w = std::max(2.0f, n_x2 - n_x1);
    float n_h = std::max(2.0f, n_y2 - n_y1);

    cv::Mat measurement(4, 1, CV_32F);
    measurement.at<float>(0) = n_cx;
    measurement.at<float>(1) = n_cy;
    measurement.at<float>(2) = n_w;
    measurement.at<float>(3) = n_h;

    kalman.correct(measurement);

    x1 = n_x1;
    y1 = n_y1;
    x2 = n_x2;
    y2 = n_y2;
    width = n_w;
    height = n_h;
    cx = n_cx;
    cy = n_cy;

    confidence = conf;
    hits++;
    missed = 0;

    if (hits >= 3) {
        confirmed = true;
    }
}

// ============================================================
// OPTICAL FLOW REFINEMENT
// ============================================================

static float compute_median(std::vector<float>& v) {
    if (v.empty()) return 0.0f;
    size_t n = v.size() / 2;
    std::nth_element(v.begin(), v.begin() + n, v.end());
    if (v.size() % 2 != 0) return v[n];
    auto max_it = std::max_element(v.begin(), v.begin() + n);
    return (*max_it + v[n]) / 2.0f;
}

bool DroneTrack::optical_flow_update(const cv::Mat& old_gray, const cv::Mat& new_gray) {
    if (old_gray.empty() || new_gray.empty()) return false;

    cv::Vec4i bbox = get_bbox();
    int h = new_gray.rows;
    int w = new_gray.cols;

    int padding = 5;
    int rx1 = std::max(0, bbox[0] - padding);
    int ry1 = std::max(0, bbox[1] - padding);
    int rx2 = std::min(w - 1, bbox[2] + padding);
    int ry2 = std::min(h - 1, bbox[3] + padding);

    if (rx2 <= rx1 || ry2 <= ry1) return false;

    cv::Mat roi = old_gray(cv::Rect(rx1, ry1, rx2 - rx1, ry2 - ry1));
    if (roi.empty()) return false;

    std::vector<cv::Point2f> points;
    cv::goodFeaturesToTrack(roi, points, 20, 0.01, 2, cv::Mat(), 3);
    if (points.empty()) return false;

    for (auto& pt : points) {
        pt.x += rx1;
        pt.y += ry1;
    }

    std::vector<cv::Point2f> new_points;
    std::vector<uchar> status;
    std::vector<float> err;

    cv::calcOpticalFlowPyrLK(
        old_gray, new_gray, points, new_points, status, err,
        cv::Size(15, 15), 3,
        cv::TermCriteria(cv::TermCriteria::EPS | cv::TermCriteria::COUNT, 20, 0.03)
    );

    std::vector<float> dx_list;
    std::vector<float> dy_list;

    for (size_t i = 0; i < status.size(); ++i) {
        if (status[i]) {
            dx_list.push_back(new_points[i].x - points[i].x);
            dy_list.push_back(new_points[i].y - points[i].y);
        }
    }

    if (dx_list.size() < 3) return false;

    float dx = compute_median(dx_list);
    float dy = compute_median(dy_list);

    cx += dx;
    cy += dy;
    x1 = cx - width / 2.0f;
    y1 = cy - height / 2.0f;
    x2 = cx + width / 2.0f;
    y2 = cy + height / 2.0f;

    return true;
}

cv::Vec4i DroneTrack::get_bbox() const {
    return cv::Vec4i((int)x1, (int)y1, (int)x2, (int)y2);
}

cv::Point2f DroneTrack::get_center() const {
    return cv::Point2f(cx, cy);
}

// ============================================================
// DRONE TRACKER CONSTRUCTOR
// ============================================================

DroneTracker::DroneTracker(int max_missing, float max_distance, float min_iou)
    : next_id(1), max_missing(max_missing), max_distance(max_distance), min_iou(min_iou) {}

void DroneTracker::reset() {
    tracks.clear();
    next_id = 1;
}

DroneTrack& DroneTracker::create_track(const cv::Vec4f& bbox, float confidence) {
    tracks.emplace_back(bbox, next_id++);
    tracks.back().confidence = confidence;
    return tracks.back();
}

// Struct for sorting cost matches
struct MatchCost {
    float cost;
    int track_idx;
    int det_idx;
};

std::vector<DroneTrack*> DroneTracker::update(const std::vector<Detection>& detections,
                                              const cv::Mat& old_gray,
                                              const cv::Mat& new_gray)
{
    // Step 1: Predict existing tracks
    std::vector<cv::Vec4i> predicted_boxes;
    for (auto& track : tracks) {
        predicted_boxes.push_back(track.predict());
    }

    // Step 2: Build possible matches
    std::vector<MatchCost> matches;

    for (size_t t_idx = 0; t_idx < tracks.size(); ++t_idx) {
        const cv::Vec4i& p_box = predicted_boxes[t_idx];
        float p_w = (float)(p_box[2] - p_box[0]);
        float p_h = (float)(p_box[3] - p_box[1]);

        float adaptive_distance = std::max(max_distance, std::max(p_w, p_h) * 3.0f);

        for (size_t d_idx = 0; d_idx < detections.size(); ++d_idx) {
            cv::Vec4f d_box(detections[d_idx].x1, detections[d_idx].y1, detections[d_idx].x2, detections[d_idx].y2);

            float iou = calculate_iou(p_box, d_box);
            float dist = center_distance(p_box, d_box);

            if (iou < min_iou && dist > adaptive_distance) {
                continue;
            }

            float distance_score = std::min(1.0f, dist / adaptive_distance);
            float cost = 0.60f * (1.0f - iou) + 0.40f * distance_score;

            matches.push_back({cost, (int)t_idx, (int)d_idx});
        }
    }

    // Step 3: Sort matches by ascending cost
    std::sort(matches.begin(), matches.end(), [](const MatchCost& a, const MatchCost& b) {
        return a.cost < b.cost;
    });

    std::vector<bool> used_tracks(tracks.size(), false);
    std::vector<bool> used_detections(detections.size(), false);
    std::vector<std::pair<int, int>> final_matches;

    for (const auto& m : matches) {
        if (used_tracks[m.track_idx] || used_detections[m.det_idx]) {
            continue;
        }
        final_matches.push_back({m.track_idx, m.det_idx});
        used_tracks[m.track_idx] = true;
        used_detections[m.det_idx] = true;
    }

    // Step 4: Update matched tracks
    for (const auto& fm : final_matches) {
        const auto& det = detections[fm.second];
        cv::Vec4f bbox(det.x1, det.y1, det.x2, det.y2);
        tracks[fm.first].update(bbox, det.confidence);
    }

    // Step 5: Optical flow for unmatched tracks
    if (!old_gray.empty() && !new_gray.empty()) {
        for (size_t i = 0; i < tracks.size(); ++i) {
            if (!used_tracks[i]) {
                tracks[i].optical_flow_update(old_gray, new_gray);
            }
        }
    }

    // Step 6: Create new tracks from unmatched detections
    for (size_t d_idx = 0; d_idx < detections.size(); ++d_idx) {
        if (!used_detections[d_idx]) {
            const auto& det = detections[d_idx];
            cv::Vec4f bbox(det.x1, det.y1, det.x2, det.y2);
            create_track(bbox, det.confidence);
        }
    }

    // Step 7: Delete lost tracks
    std::vector<DroneTrack> remaining_tracks;
    for (const auto& track : tracks) {
        if (track.missed <= max_missing) {
            remaining_tracks.push_back(track);
        }
    }
    tracks = std::move(remaining_tracks);

    // Return confirmed active tracks
    std::vector<DroneTrack*> active_tracks;
    for (auto& track : tracks) {
        if (track.confirmed) {
            active_tracks.push_back(&track);
        }
    }

    return active_tracks;
}

File: deployment_custom_tracker/drone_tracker.py
import cv2
import numpy as np
import math


class DroneTrack:

    def __init__(self, bbox, track_id):

        self.id = track_id

        x1, y1, x2, y2 = bbox

        self.x1 = float(x1)
        self.y1 = float(y1)
        self.x2 = float(x2)
        self.y2 = float(y2)

        self.width = max(2.0, self.x2 - self.x1)
        self.height = max(2.0, self.y2 - self.y1)

        self.cx = (self.x1 + self.x2) / 2
        self.cy = (self.y1 + self.y2) / 2

        # ====================================================
        # KALMAN FILTER
        #
        # State:
        #
        # x
        # y
        # vx
        # vy
        # width
        # height
        # ====================================================

        self.kalman = cv2.KalmanFilter(6, 4)

        self.kalman.transitionMatrix = np.array(
            [
                [1, 0, 1, 0, 0, 0],
                [0, 1, 0, 1, 0, 0],
                [0, 0, 1, 0, 0, 0],
                [0, 0, 0, 1, 0, 0],
                [0, 0, 0, 0, 1, 0],
                [0, 0, 0, 0, 0, 1]
            ],
            dtype=np.float32
        )

        self.kalman.measurementMatrix = np.array(
            [
                [1, 0, 0, 0, 0, 0],
                [0, 1, 0, 0, 0, 0],
                [0, 0, 0, 0, 1, 0],
                [0, 0, 0, 0, 0, 1]
            ],
            dtype=np.float32
        )

        # Process noise
        self.kalman.processNoiseCov = np.eye(
            6,
            dtype=np.float32
        ) * 0.03

        # Measurement noise
        self.kalman.measurementNoiseCov = np.eye(
            4,
            dtype=np.float32
        ) * 0.15

        self.kalman.errorCovPost = np.eye(
            6,
            dtype=np.float32
        )

        self.kalman.statePost = np.array(
            [
                [self.cx],
                [self.cy],
                [0],
                [0],
                [self.width],
                [self.height]
            ],
            dtype=np.float32
        )

        # ====================================================
        # TRACK STATE
        # ====================================================

        self.age = 1

        self.hits = 1

        self.missed = 0

        self.confirmed = False

        self.confidence = 0.0

        self.history = []

        self.last_gray = None

    # ========================================================
    # PREDICT
    # ========================================================

    def predict(self):

        prediction = self.kalman.predict()

        self.cx = float(prediction[0][0])
        self.cy = float(prediction[1][0])

        vx = float(prediction[2][0])
        vy = float(prediction[3][0])

        self.width = max(
            2.0,
            float(prediction[4][0])
        )

        self.height = max(
            2.0,
            float(prediction[5][0])
        )

        self.x1 = self.cx - self.width / 2
        self.y1 = self.cy - self.height / 2

        self.x2 = self.cx + self.width / 2
        self.y2 = self.cy + self.height / 2

        self.age += 1

        self.missed += 1

        return self.get_bbox()

    # ========================================================
    # UPDATE FROM YOLO
    # ========================================================

    def update(self, bbox, confidence):

        x1, y1, x2, y2 = bbox

        cx = (x1 + x2) / 2
        cy = (y1 + y2) / 2

        width = max(
            2.0,
            x2 - x1
        )

        height = max(
            2.0,
            y2 - y1
        )

        measurement = np.array(
            [
                [cx],
                [cy],
                [width],
                [height]
            ],
            dtype=np.float32
        )

        self.kalman.correct(
            measurement
        )

        self.x1 = float(x1)
        self.y1 = float(y1)
        self.x2 = float(x2)
        self.y2 = float(y2)

        self.width = width
        self.height = height

        self.cx = cx
        self.cy = cy

        self.confidence = confidence

        self.hits += 1

        self.missed = 0

        self.history.append(
            (cx, cy)
        )

        if len(self.history) > 30:
            self.history.pop(0)

        # Confirm after several successful updates

        if self.hits >= 3:
            self.confirmed = True

    # ========================================================
    # OPTICAL FLOW REFINEMENT
    # ========================================================

    def optical_flow_update(
        self,
        old_gray,
        new_gray
    ):

        if old_gray is None:
            return False

        x1, y1, x2, y2 = self.get_bbox()

        h, w = new_gray.shape

        # Expand region slightly

        padding = 5

        x1 = max(
            0,
            x1 - padding
        )

        y1 = max(
            0,
            y1 - padding
        )

        x2 = min(
            w - 1,
            x2 + padding
        )

        y2 = min(
            h - 1,
            y2 + padding
        )

        if x2 <= x1 or y2 <= y1:
            return False

        # ----------------------------------------------------
        # Create feature points around drone
        # ----------------------------------------------------

        roi = old_gray[
            y1:y2,
            x1:x2
        ]

        if roi.size == 0:
            return False

        points = cv2.goodFeaturesToTrack(
            roi,
            maxCorners=20,
            qualityLevel=0.01,
            minDistance=2,
            blockSize=3
        )

        if points is None:
            return False

        points[:, :, 0] += x1
        points[:, :, 1] += y1

        # ----------------------------------------------------
        # Optical flow
        # ----------------------------------------------------

        new_points, status, error = cv2.calcOpticalFlowPyrLK(
            old_gray,
            new_gray,
            points,
            None,
            winSize=(15, 15),
            maxLevel=3,
            criteria=(
                cv2.TERM_CRITERIA_EPS |
                cv2.TERM_CRITERIA_COUNT,
                20,
                0.03
            )
        )

        if new_points is None:
            return False

        good_old = points[
            status == 1
        ]

        good_new = new_points[
            status == 1
        ]

        if len(good_new) < 3:
            return False

        # ----------------------------------------------------
        # Calculate movement
        # ----------------------------------------------------

        movement = good_new - good_old

        dx = np.median(
            movement[:, 0]
        )

        dy = np.median(
            movement[:, 1]
        )

        # ----------------------------------------------------
        # Update center
        # ----------------------------------------------------

        self.cx += dx
        self.cy += dy

        self.x1 = (
            self.cx -
            self.width / 2
        )

        self.y1 = (
            self.cy -
            self.height / 2
        )

        self.x2 = (
            self.cx +
            self.width / 2
        )

        self.y2 = (
            self.cy +
            self.height / 2
        )

        return True

    # ========================================================
    # BOUNDING BOX
    # ========================================================

    def get_bbox(self):

        return (
            int(self.x1),
            int(self.y1),
            int(self.x2),
            int(self.y2)
        )

    # ========================================================
    # CENTER
    # ========================================================

    def get_center(self):

        return (
            float(self.cx),
            float(self.cy)
        )

    # ========================================================
    # VELOCITY
    # ========================================================

    def get_velocity(self):

        state = self.kalman.statePost

        return (
            float(state[2][0]),
            float(state[3][0])
        )


# ============================================================
# IOU
# ============================================================

def calculate_iou(box1, box2):

    x1 = max(
        box1[0],
        box2[0]
    )

    y1 = max(
        box1[1],
        box2[1]
    )

    x2 = min(
        box1[2],
        box2[2]
    )

    y2 = min(
        box1[3],
        box2[3]
    )

    intersection_width = max(
        0,
        x2 - x1
    )

    intersection_height = max(
        0,
        y2 - y1
    )

    intersection = (
        intersection_width *
        intersection_height
    )

    area1 = (
        max(
            0,
            box1[2] - box1[0]
        )
        *
        max(
            0,
            box1[3] - box1[1]
        )
    )

    area2 = (
        max(
            0,
            box2[2] - box2[0]
        )
        *
        max(
            0,
            box2[3] - box2[1]
        )
    )

    union = (
        area1 +
        area2 -
        intersection
    )

    if union <= 0:
        return 0.0

    return intersection / union


# ============================================================
# CENTER DISTANCE
# ============================================================

def center_distance(box1, box2):

    c1x = (
        box1[0] +
        box1[2]
    ) / 2

    c1y = (
        box1[1] +
        box1[3]
    ) / 2

    c2x = (
        box2[0] +
        box2[2]
    ) / 2

    c2y = (
        box2[1] +
        box2[3]
    ) / 2

    return math.sqrt(
        (c1x - c2x) ** 2 +
        (c1y - c2y) ** 2
    )


# ============================================================
# DRONE TRACKER
# ============================================================

class DroneTracker:

    def __init__(
        self,

        max_missing=10,

        max_distance=100,

        min_iou=0.01
    ):

        self.tracks = []

        self.next_id = 1

        self.max_missing = max_missing

        self.max_distance = max_distance

        self.min_iou = min_iou

    # ========================================================
    # START TRACKING
    # ========================================================

    def create_track(
        self,
        bbox,
        confidence
    ):

        track = DroneTrack(
            bbox,
            self.next_id
        )

        track.confidence = confidence

        self.next_id += 1

        self.tracks.append(
            track
        )

        return track

    # ========================================================
    # UPDATE
    # ========================================================

    def update(
        self,
        detections,
        old_gray=None,
        new_gray=None
    ):

        # ====================================================
        # STEP 1
        # Predict existing tracks
        # ====================================================

        predicted_boxes = []

        for track in self.tracks:

            predicted_boxes.append(
                track.predict()
            )

        # ====================================================
        # STEP 2
        # Build possible matches
        # ====================================================

        matches = []

        for track_index, track in enumerate(
            self.tracks
        ):

            predicted_box = predicted_boxes[
                track_index
            ]

            for detection_index, detection in enumerate(
                detections
            ):

                detection_box = detection[
                    :4
                ]

                iou = calculate_iou(
                    predicted_box,
                    detection_box
                )

                distance = center_distance(
                    predicted_box,
                    detection_box
                )

                # --------------------------------------------
                # Adaptive distance
                # --------------------------------------------

                box_width = (
                    predicted_box[2] -
                    predicted_box[0]
                )

                box_height = (
                    predicted_box[3] -
                    predicted_box[1]
                )

                adaptive_distance = max(
                    self.max_distance,
                    max(
                        box_width,
                        box_height
                    ) * 3
                )

                # --------------------------------------------
                # Reject impossible matches
                # --------------------------------------------

                if (
                    iou < self.min_iou
                    and
                    distance > adaptive_distance
                ):
                    continue

                # --------------------------------------------
                # Normalize distance
                # --------------------------------------------

                distance_score = min(
                    1.0,
                    distance /
                    adaptive_distance
                )

                # --------------------------------------------
                # Cost
                #
                # Lower is better
                # --------------------------------------------

                cost = (

                    0.60 *
                    (1.0 - iou)

                    +

                    0.40 *
                    distance_score

                )

                matches.append(
                    (
                        cost,
                        track_index,
                        detection_index
                    )
                )

        # ====================================================
        # STEP 3
        # Sort matches
        # ====================================================

        matches.sort(
            key=lambda x: x[0]
        )

        used_tracks = set()

        used_detections = set()

        final_matches = []

        for (
            cost,
            track_index,
            detection_index
        ) in matches:

            if track_index in used_tracks:
                continue

            if detection_index in used_detections:
                continue

            final_matches.append(
                (
                    track_index,
                    detection_index
                )
            )

            used_tracks.add(
                track_index
            )

            used_detections.add(
                detection_index
            )

        # ====================================================
        # STEP 4
        # Update matched tracks
        # ====================================================

        for (
            track_index,
            detection_index
        ) in final_matches:

            detection = detections[
                detection_index
            ]

            bbox = detection[
                :4
            ]

            confidence = detection[
                4
            ]

            self.tracks[
                track_index
            ].update(
                bbox,
                confidence
            )

        # ====================================================
        # STEP 5
        # Optical flow for tracks that were not detected
        # ====================================================

        for index, track in enumerate(
            self.tracks
        ):

            if index in used_tracks:
                continue

            if (
                old_gray is not None
                and
                new_gray is not None
            ):

                track.optical_flow_update(
                    old_gray,
                    new_gray
                )

        # ====================================================
        # STEP 6
        # Create new tracks from unmatched detections
        # ====================================================

        for detection_index, detection in enumerate(
            detections
        ):

            if detection_index in used_detections:
                continue

            self.create_track(
                detection[:4],
                detection[4]
            )

        # ====================================================
        # STEP 7
        # Delete lost tracks
        # ====================================================

        remaining_tracks = []

        for track in self.tracks:

            if track.missed <= self.max_missing:

                remaining_tracks.append(
                    track
                )

        self.tracks = remaining_tracks

        # ====================================================
        # Return active tracks
        # ====================================================

        return [
            track

            for track in self.tracks

            if track.confirmed
        ]

    # ========================================================
    # CLEAR
    # ========================================================

    def reset(self):

        self.tracks = []

        self.next_id = 1